### PR TITLE
add safe libc::clock_nanosleep wrapper

### LINKED ISSUE #1299: Add wrappers for some time-related syscalls
`gettimeofday`, ~`clock_gettime`~, ~`clock_settime`~, ~`clock_getres`~, `nanosleep` and `clock_nanosleep` seem to be supported by various *nixes (at least Linux and FreeBSD), but bindings are not included in this crate.

I'd be happy to open a PR adding these bindings.

### PR DESCRIPTION
**Pull request by maxbla**
Fixes #1299

### CODE REVIEW

File: src/time.rs
Comment (maxbla) [inline src/time.rs:260]: I'm curious what maintainers think of this type alias. I worry it might not be forwards compatible.

Comment (maxbla) [reply src/time.rs:260]: Never mind - that worked by chance, only because libc::TIMER_ABSTIME == 1 and libc::TFD_TIMER_ABSTIME == 1. I fixed it by creating an entirely new type.

**Comment by maxbla** [comment]
`libc::clock_nanosleep` isn't defined for all platforms. https://github.com/rust-lang/libc/pull/1922 tries to fix this.

**Comment by maxbla** [comment]
this PR is waiting on the next minor release of libc

**Comment by kamalmarhubi** [comment]
@maxbla
> this PR is waiting on the next minor release of libc

looks like libc 0.2.80 was cut a couple of weeks ago, and includes https://github.com/rust-lang/libc/pull/1922. could you bump the dependency version in `Cargo.toml`?

**Comment by jluebbe** [comment]
> @maxbla
>
> looks like libc 0.2.80 was cut a couple of weeks ago, and includes [rust-lang/libc#1922](https://github.com/rust-lang/libc/pull/1922). could you bump the dependency version in `Cargo.toml`?

It seems to me that the bump is already part of the PR:
https://github.com/nix-rust/nix/pull/1315/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R20

**Comment by maxbla** [comment]
The new problem with this PR is that `libc` added `clock_nanosleep` for FreeBSD 12+ only. `clock_nanosleep` was implemented in FreeBSD 11.1, so `libc` couldn't add a wrapper for FreeBSD 11 (if it did, `libc` would fail to build for FreeBSD 11.0). As long as FreeBSD 11.x remains in tier 1 for `nix`, this PR can't be merged.

**Comment by asomers** [comment]
> The new problem with this PR is that `libc` added `clock_nanosleep` for FreeBSD 12+ only. `clock_nanosleep` was implemented in FreeBSD 11.1, so `libc` couldn't add a wrapper for FreeBSD 11 (if it did, `libc` would fail to build for FreeBSD 11.0). As long as FreeBSD 11.x remains in tier 1 for `nix`, this PR can't be merged.

Actually, that's not a problem.  For one thing, Rust's FFI bindings don't know whether a function is present at build time.  And they don't care whether it's present at runtime unless somebody tries to use it.  For another thing, FreeBSD 11.0 is EoL.  We shouldn't worry about it anymore.  Long Live FreeBSD 11.4.

**Comment by wolthom** [comment]
Are there any updates to this?

**Comment by marcfir** [comment]
Is it realistic that there will be progress here?

**Comment by SteveLauC** [comment]
Gentle ping on the author @maxbla, are you still interested in finishing this PR? If not, I think I will pick it up add it to Nix:)

File: src/time.rs
Comment (SteveLauC) [inline src/time.rs:266]: We should use the `libc_flags` macro instead:

```suggestion
libc_bitflags! {
    /// Flags that are used for arming the timer.
    pub struct ClockNanosleepFlags: libc::c_int {
        TIMER_ABSTIME;
    }
}
```

File: src/time.rs
Comment (SteveLauC) [inline src/time.rs:282]: We use cfg aliases now:

```suggestion
#[cfg(any(
    linux_android,
    solarish,
    target_os = "freebsd",
    target_os = "netbsd"
))]
```

File: src/time.rs
Comment (SteveLauC) [inline src/time.rs:300]:
```suggestion
        Err(Errno::from_i32(ret))
```

**Review by SteveLauC — CHANGES_REQUESTED** [review]
I have left some comments on what needs to be changed if you want to finish this, and a CHANGELOG entry is needed, please see [CONTRIBUTING.md](https://github.com/nix-rust/nix/blob/master/CONTRIBUTING.md#pull-requests) on how to add one.

File: src/time.rs
Comment (SteveLauC) [reply src/time.rs:266]: And, adding a doc comment for this flag would be great:)

**Pull request closed without merging**
